src/rust/engine/src/rule_graph.rs

```diff
@@ -475,8 +451,80 @@ impl <'a> GraphMaker<'a> {
 
   fn _remove_unfulfillable_rules_and_dependents(&self,
                                                 rule_graph: RuleGraph) -> RuleGraph {
-    // TODO port this portion
-    rule_graph
+    // Removes all unfulfillable rules transitively from the roots and the dependency edges.
+    //
+    // Takes the current root rule set and dependency table and removes all rules that are not
+    // transitively fulfillable.
+    //
+    // Deforestation. Leaping from tree to tree.
+
+    // could experiment with doing this for each rule added and deduping the traversal list
+    let mut new_unfulfillable_rules = rule_graph.unfulfillable_rules.clone();
+    let mut new_dependency_edges = rule_graph.rule_dependency_edges.clone();
+    let mut new_root_dependencies = rule_graph.root_dependencies.clone();
+
+    fn update_edges_based_on_unfulfillable_entry<K>(edge_container: &mut HashMap<K, RuleEdges>,
```

Comment:
I'd initially started with porting the python directly, and it has a number of them. When it worked I went with it.

Since I needed to explicitly pass most of the closed over things, there's no real reason why they need to be inner anymore. I'll drop them into the outer scope.